### DISCUSSION TITLE
Is the default strength of DRBG in 3.0 FIPS provider - 1024 or 256 bits?

### DISCUSSION THREAD

**Discussion by visweshn92** [openssl/openssl#21758] [Q&A] [answered]
Assuming we build the OpenSSL3 FIPS provider with the default seeding option (i.e. OS seeding), what is the value of DRBG strength? 

The code appears to pick 1024 bits (from 81aef6ba720971c09ad68f89d418c8d3d3f904f7 of @paulidale 

providers/implementations/rands/seed_src.c
```
static int seed_src_get_ctx_params(void *vseed, OSSL_PARAM params[])
{
    PROV_SEED_SRC *s = (PROV_SEED_SRC *)vseed;
    OSSL_PARAM *p;

    p = OSSL_PARAM_locate(params, OSSL_RAND_PARAM_STATE);
    if (p != NULL && !OSSL_PARAM_set_int(p, s->state))
        return 0;

    p = OSSL_PARAM_locate(params, OSSL_RAND_PARAM_STRENGTH);
    if (p != NULL && !OSSL_PARAM_set_int(p, 1024))          ====================> HERE
        return 0;

    p = OSSL_PARAM_locate(params, OSSL_RAND_PARAM_MAX_REQUEST);
    if (p != NULL && !OSSL_PARAM_set_size_t(p, 128))
        return 0;
    return 1;
}

```

But the documentation appears to state 256 bits. 

https://www.openssl.org/docs/man3.0/man1/openssl-rand.html
```
The random bytes are generated using the [RAND_bytes(3)](https://www.openssl.org/docs/man3.0/man3/RAND_bytes.html) function, which provides a security level of 256 bits,
```
https://www.openssl.org/docs/man3.0/man3/RAND_bytes.html
```
By default, the OpenSSL CSPRNG supports a security level of 256 bits, provided it was able to seed itself from a trusted entropy source. On all major platforms supported by OpenSSL (including the Unix-like platforms and Windows), OpenSSL is configured to automatically seed the CSPRNG on first use using the operating systems's random generator.
```

Can anyone confirm the documentation (or) point me to the correct code logic?

**Comment by t8m** [comment, marked as answer]
The code above provides strength of the seed source,  not the DRBG itself. The number 1024 is arbitrary to not limit the strength of the whole chain. The default DRBG type fetched is the AES-256-CTR DRBG which should provide strength of 256 bits. The resulting strength of the whole chain should be the minimum of all the strengths of RBGs in the chain which is 256 bits.

**Comment by t8m** [comment]
BTW the seed source is outside the FIPS module boundary. The FIPS provider depends on external seeding for its DRBG.

**Reply by visweshn92** [reply]
Yes, I wanted to know how many bytes/bits by default OpenSSL DRBG fetches from the OS during each seeding.  

My reading of the code is that value is ultimately coming from the "strength" field of drbg (assuming it is within configured min_entropylen and max_entroylen) which was initially set as int value 1024 as shown in the snippet.

**Reply by t8m** [reply]
It fetches 48 bytes by default.

**Reply by t8m** [reply]
The number 1024 in the code above does NOT apply to the DRBG strength. Strength of the CTR drbg is set here:
https://github.com/openssl/openssl/blob/master/providers/implementations/rands/drbg_ctr.c#L584

**Reply by t8m** [reply]
On the next line of the code you see that the seed length - that is what is fetched from OS - is keylen + 16 = 32 + 16 = 48 bytes.

**Reply by visweshn92** [reply]
Thank you for the pointer on where exactly the strength is set and also for `drbg->seedlen = keylen + 16;` explains the 48 bytes (i.e. 32 + 16) fetch from OS. 


Somehow the int values in `seed_src.c` misled me to think that is what dictates about seed sources (which by default OpenSSL picks OS entropy).

**Reply by paulidale** [reply]
SP 800-90Ar1 section 9.1 indicates that it is okay to increase the entropy required by 50% and not have a separate nonce.  We do this from memory:
https://github.com/openssl/openssl/blob/60421893a286bb9eb7fb7c2454b84af9778ffca4/providers/implementations/rands/drbg.c#L413-L422

The best way to check would be to strace and see how much is read.

**Reply by harshbhavsar30** [reply]
As Key Length is 32, we receive 32+16=48 as minimum entropy length. I need to change minimum entropy length.
**Is it possible to set minimum and maximum entropy length for DRBG context? If Yes, please share the possible way to accomplish it.**

**Reply by paulidale** [reply]
It's not, the amount of entropy required is defined by the standards.
The extra 16 bytes are for the nonce as mentioned.
